### PR TITLE
Short circuiting and lazy evaluation of CSG regions for cell searching

### PR DESCRIPTION
**Pull request by myerspat**
In order to reduce the execution time of `contains_complex` for cell searching, the algorithm was changed to accommodate short-circuiting and lazy evaluation of surface half-spaces. The function now uses a CSG region in infix notation rather than postfix. To ensure precedence is parentheses, complements, intersections, and unions, a function, `add_precedence`, was introduced. Furthermore, all complements were removed using De Morgan's laws in `remove_complement_ops`. The new algorithm is as follows:
1. Two variables are initialized: a boolean, `in_cell`, of whether the particle is in the cell or not, and an integer, `total_depth`, is the depth of parenthesis at the current iterator and is initially set to zero
2. For each token in the CSG region: if the token is a surface then its half-space is evaluated, if it is an intersection and `in_cell` is false or the token is a union and `in_cell` is true the region is short-circuited using the `total_depth` and relative depth, if the token is a left parenthesis `total_depth` is incremented, and if the token is a right parenthesis `total_depth` is decremented

### CODE REVIEW

**Comment by paulromano** [comment]
@myerspat It looks like `test_cell_box` in `tests/unit_tests/test_complex_cell_bb.py` is segfaulting. Right now the algorithm still assumes that the argument being passed in is in postfix notation and we are passing it infix. An easy, although perhaps inefficient, fix is to generate the postfix expression at the time `CSGCell::bounding_box()` is called and then pass it to `bounding_box_complex`. Otherwise you'll have to write an algorithm that evaluates the bounding boxes based on the infix expression.

File: src/cell.cpp [hunk 1]
```diff
@@ -107,13 +107,105 @@ vector<int32_t> tokenize(const std::string region_spec)
...
+          return_iterator = start - 1;
+        }
+        return return_iterator;
+      } else {
+        start = infix.insert(start, OP_RIGHT_PAREN);
+        if (return_iterator == infix.begin()) {
+          return_iterator = start - 1;
+        }
+        return return_iterator;
+      }
```
Comment (paulromano) [inline src/cell.cpp:165]: These could be combined to simplify a bit:
```suggestion
      } else {
        start = infix.insert(
          start_token == OP_UNION ? start - 1 : start, OP_RIGHT_PAREN);
        if (return_iterator == infix.begin()) {
          return_iterator = start - 1;
        }
        return return_iterator;
      }
```

File: src/cell.cpp [hunk 2]
```diff
@@ -107,13 +107,105 @@ vector<int32_t> tokenize(const std::string region_spec)
   return tokens;
 }
 
+//==============================================================================
+//! Add precedence for infix regions so intersections have higher
+//! precedence than unions using parenthesis.
+//==============================================================================
+
+std::vector<int32_t>::iterator add_parenthesis(
```
Comment (paulromano) [inline src/cell.cpp:115]:
```suggestion
std::vector<int32_t>::iterator add_parentheses(
```

File: src/cell.cpp [hunk 3]
```diff
@@ -107,13 +107,105 @@ vector<int32_t> tokenize(const std::string region_spec)
...
+    // If the token is a union another operator has not been found set the
+    // current operator to union
+    // If the current operator is a union and the token is an intersection
+    // assert precedence
+    // If the token is a parenthesis reset the current operator
+    if (token == OP_UNION || token == OP_INTERSECTION) {
+      if (current_op == 0) {
+        current_op = token;
+      } else if (token != current_op) {
+        it = add_parenthesis(it, infix);
```
Comment (paulromano) [inline src/cell.cpp:193]:
```suggestion
        it = add_parentheses(it, infix);
```

File: src/cell.cpp [hunk 4]
```diff
@@ -107,13 +107,105 @@ vector<int32_t> tokenize(const std::string region_spec)
...
+  // Add right parenthesis
+  // While the start iterator is within the bounds of infix
+  while (start < infix.end()) {
+    start++;
+
+    // If the current token is an operator and is different than the start token
+    if (*start >= OP_UNION && *start != start_token) {
+      // Skip wraped regions but save iterator position to check precedence and
+      // add right parenthesis depending on the precedence of the original token
+      // when a right parenthesis is encountered of the opposite token
```
Comment (paulromano) [inline src/cell.cpp:139]: I'm confused by "is encountered of" -- can you reword this?

File: src/cell.cpp [hunk 5]
```diff
@@ -107,13 +107,105 @@ vector<int32_t> tokenize(const std::string region_spec)
   return tokens;
 }
 
+//==============================================================================
+//! Add precedence for infix regions so intersections have higher
+//! precedence than unions using parenthesis.
```
Comment (paulromano) [inline src/cell.cpp:112]:
```suggestion
//! precedence than unions using parentheses.
```

File: src/cell.cpp [hunk 6]
```diff
@@ -616,19 +705,18 @@ void CSGCell::to_hdf5_inner(hid_t group_id) const
...
         auto surf_id = model::surfaces[abs(token) - 1]->id_;
         region_spec << " " << ((token > 0) ? surf_id : -surf_id);
+      } else if (token > OP_COMPLEMENT) {
+        if (token == OP_LEFT_PAREN) {
+          region_spec << " (";
+        } else {
+          region_spec << " )";
+        }
+      } else if (token == OP_UNION) {
+        region_spec << " |";
```
Comment (paulromano) [inline src/cell.cpp:719]: Are these changes really needed? It looks to me like the previous code would have worked fine?

File: src/cell.cpp [hunk 7]
```diff
@@ -656,16 +744,16 @@ void CSGCell::apply_demorgan(
       *start = OP_UNION;
     }
     start++;
-  }
+  } while (start < stop);
```
Comment (paulromano) [inline src/cell.cpp:747]: Why was the change from a `while` to a `do/while` loop necessary?

**Review by paulromano — CHANGES_REQUESTED** [review]
Very nice work on this @myerspat! A few questions and suggestions:

File: src/cell.cpp [hunk 8]
```diff
@@ -107,13 +107,105 @@ vector<int32_t> tokenize(const std::string region_spec)
...
+std::vector<int32_t>::iterator add_parenthesis(
+  std::vector<int32_t>::iterator start, std::vector<int32_t>& infix)
+{
+  int32_t start_token = *start;
+  // Add left parenthesis
+  if (start_token == OP_INTERSECTION) {
+    start = infix.insert(start - 1, OP_LEFT_PAREN);
+  } else {
+    start = infix.insert(start + 1, OP_LEFT_PAREN);
+  }
```
[the remark this answers is not in the document]
Comment (myerspat) [reply src/cell.cpp:124]: Intersections take precedence over unions, If we have a | b & c and our current token is the intersection then we want to wrap the b and c like a | (b & c), that's where the `start-1` comes from for the left parenthesis. In the case of unions we don't want to include the surface that is connected to the intersection so we have `start+1`.

File: src/cell.cpp (hunk 6 above)
Comment (myerspat) [reply src/cell.cpp:719]: No, they are not, I changed it when I was debugging. We can get rid of the complement if statement as we've already done De Morgan's laws.

Comment (paulromano) [reply src/cell.cpp:714]: Yeah, let's just revert to the previous logic here. I would actually just keep the `if` block for complement since then the `else` block at the bottom definitely refers to a non-operator.

**Review by paulromano — APPROVED** [review]
All looks good now. Thanks so much @myerspat for this great improvement. One more check that I'll mention I did: I ran simulations of the ITER E-lite model with a mesh tally and compared this branch and develop and I'm getting consistent results.
